### ISSUE TITLE
'dim.' vs 'dim'

### ISSUE THREAD

**Issue by hpages** [HenrikBengtsson/matrixStats#163]
This error message is not accurate:
```
> colRanges(array(1:24, 4:2))
Error in colRanges(array(1:24, 4:2)) : 
  Argument 'dim' must be an integer vector of length two.
```
The name of the argument is `dim.`, not `dim`.

BTW, why is the argument `dim.` and not `dim`, like in the `array()` constructor?

**Comment by HenrikBengtsson** [comment]
Thanks for reporting.  The use `dim` in the error message is probably a leftover from the past where I/we aimed to use `dim` instead of `dim.` (see below).  This has now been fixed in commit 286d698.

> BTW, why is the argument `dim.` and not `dim`, like in the `array()` constructor?

It took me a while to recover the reason for naming that odd argument name `dim.` and not `dim` but I knew there was a reason for it.  The clue is in the commit 5a5dad75 (2014-11-24) message:

> Now using argument 'dim.=dim(x)', cf. https://stat.ethz.ch/pipermail/r-devel/2014-November/070155.html

That post (of mine) points out that we cannot use `dim=dim(x)` because then we get:

```r
foo <- function(x, dim=dim(x)) { dim }
> foo(matrix(0, 2, 3))
Error in foo(matrix(0, 2, 3)) : 
  promise already under evaluation: recursive default argument
reference or earlier problems?
```

I kinda (re-)asked myself the question whether `dim.` is a good name in https://github.com/HenrikBengtsson/matrixStats/issues/83;

> Also, should it be called `dim.` or just `dim`?

I agree - it's an "odd" choice of argument name.  Re-reading my R-devel thread, maybe the cleanest yet perfomant solution would be to use:
```r
foo <- function(x, dim=NULL) {
  if (is.null(dim)) dim <- dim(x)
  dim 
}
```
and then document what `dim = NULL` resolves to.
